libbpf-tools/trace_helpers.h

```diff
@@ -0,0 +1,14 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef __TRACE_HELPER_H
+#define __TRACE_HELPER_H
+
+struct ksym {
+	long addr;
+	char *name;
+};
+
+int load_kallsyms(void);
+struct ksym *ksym_search(long key);
```

Comment:
let's agree on this API and then, once your PR is merged, I'll go and optimize internals a bit. It will be easier to coordinate (and test for me) that way. API that I think should be good is to avoid global state, so how about this:

```c
struct ksyms {
    struct ksym syms*;
    int sym_cnt;
};

struct ksyms *ksyms__load(void);
void ksyms__free(struct ksyms *ksyms);
const struct ksym *ksyms__map_addr(const struct ksyms *ksyms, long addr);
const struct ksym *ksyms__get_symbol(const struct ksyms *ksyms, const char *name);
```

Application then will have to:

```c
struct ksyms *syms = ksyms__load();
if (!syms) /* error */

struct ksym *sym = ksyms__map_addr(syms, 123);
if (!sym) /* not found */

/* once done, free ksyms cache */
ksysms__free(syms);
```

Does this API make sense?

Reply:
Agree, thanks.

Reply:
Hi @anakryiko , I've implemented the API and updated the PR, but changed a little, looks like:

- The `trace_helpers.h`

``` c

struct ksym {
 	long addr;
 	char name[0];
};

typedef struct ksyms ksyms_t;

ksyms_t *ksyms__load(void);
void ksyms__free(ksyms_t *ksyms);
const struct ksym *ksyms__map_addr(const ksyms_t *ksyms, long addr);
const struct ksym *ksyms__get_symbol(const ksyms_t *ksyms, const char *name);
```

I ran `cat /proc/kallsyms| wc -l` on` kernel 5.6 rc2` got `144455`. So I think maybe we need to use `rbtree` to get efficient query efficiency, right?

So I defined 

- The `trace_helpers.c` with:

``` c
struct ksym_rb_nodes {
        struct rb_node addr_rb_node;
        struct rb_node name_rb_node;
        struct ksym ksym;
};

struct ksyms {
        struct rb_root addr_entries;
        struct rb_root name_entries;
};
```

The steps are:
1. Read `/proc/kallsyms` line by line and create `addrs_entries` rbtree;
2. After `adds_rb_tree` has completed, use it to gen `names_entries` entree;

So `ksyms__map_addr` visits `addrs_entries` and `ksyms__get_symbol` visits `names_entries` to return `ksym`.

It seems different from your thought:

> given this is going to be potentially used by many tools (kallsyms are also useful for kernel stack symbolization), I think it will pay off to do this a bit more efficiently and cleanly:
> 
> ```
> struct kallsyms_cache {
>     size_t nr_syms;
>     struct ksym *syms;
>     char *strs;
> };
> ```
> 
> `strs` in this case would be a contiguous piece of memory containing all the strings that are needed for all read symbols (this will avoid a lot of small memory allocations performed by strdup). Then each ksym's name pointer will just point somewhere into this piece of memory. I can prototype something like this quickly over weekend, probably, if you are interested.

What do you think about it? Does this implement make sense? Thank you.